File: algorithms/python/minimumIndexSumofTwoLists/minimumIndexSumofTwoLists.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/minimum-index-sum-of-two-lists/
# Author: Miao Zhang
# Date:   2021-02-24

class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        d1, d2 = ({r: i for (i, r) in enumerate(l)} for l in (list1, list2))
        c = set(list1).intersection(list2)
        m = min(d1[x] + d2[x] for x in c)
        return [x for x in c if m == d1[x] + d2[x]]

File: algorithms/cpp/minimumIndexSumofTwoLists/minimumIndexSumofTwoLists.cpp
// Source: https://leetcode.com/problems/minimum-index-sum-of-two-lists/
// Author: Miao Zhang
// Date:   2021-02-24

class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        unordered_map<string, int> dic;
        for (int i = 0; i < list1.size(); i++) {
            dic[list1[i]] = i;
        }
        vector<string> res;
        int bestsums = INT_MAX;
        for (int i = 0; i < list2.size(); i++) {
            if (!dic.count(list2[i])) continue;
            int sums = dic[list2[i]] + i;
            if (sums < bestsums) res.clear();
            if (sums <= bestsums) {
                bestsums = sums;
                res.push_back(list2[i]);
            }
        }
        return res;
    }
};
